hathor/indexes/addresses_rocksdb_index.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2021 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Iterable, List, Optional, Set, Tuple
+
+from structlog import get_logger
+
+from hathor.indexes.addresses_index import AddressesIndex
+from hathor.pubsub import HathorEvents
+from hathor.transaction import BaseTransaction
+from hathor.transaction.scripts import parse_address_script
+
+if TYPE_CHECKING:  # pragma: no cover
+    import rocksdb
+
+    from hathor.pubsub import EventArguments, PubSubManager
+    from hathor.transaction import TxOutput
+
+logger = get_logger()
+
+_CF_NAME_ADDRESS_INDEX = b'address-index'
+
+
+class AddressesRocksDBIndex(AddressesIndex):
+    """ Index of inputs/outputs by address.
+
+    This index uses rocksdb and the following key format:
+
+        key = [address][tx.timestamp][tx.hash]
+              |--34b--||--4 bytes---||--32b--|
+
+    It works nicely because rocksdb uses a tree sorted by key under the hoods.
+
+    The timestamp must be serialized in big-endian, so ts1 > ts2 implies that bytes(ts1) > bytes(ts2),
+    hence the transactions are sorted by timestamp.
+    """
+    def __init__(self, db: 'rocksdb.DB', *, cf_name: Optional[bytes] = None,
+                 pubsub: Optional['PubSubManager'] = None) -> None:
+        self._db = db
+        self._cf_name = cf_name or _CF_NAME_ADDRESS_INDEX
+        self._cf = self._db.get_column_family(self._cf_name)
+        self.pubsub = pubsub
+        if self.pubsub:
+            self.subscribe_pubsub_events()
+
+    def _to_key(self, address: str, tx: Optional[BaseTransaction] = None) -> bytes:
+        import struct
+        assert len(address) == 34
+        key = address.encode('ascii')
+        if tx:
+            assert tx.hash is not None
+            assert len(tx.hash) == 32
+            key += struct.pack('>I', tx.timestamp) + tx.hash
+            assert len(key) == 34 + 4 + 32
+        return key
+
+    def _from_key(self, key: bytes) -> Tuple[str, int, bytes]:
```

Comment:
I thought about turning this return a namedtuple but given that we use only once this structure (and we get it twice), I think it's fine to keep it like this. But if we start using the same tuple in more places, having a namedtuple would be better in my opinion

Reply:
I think it's easy enough to make it a namedtuple and will improve readability and make serialization/deserialization stuff closer. I'll do that.

Reply:
While making this I realized it's a bit awkward because of the asymmetry of serialize/deserialize (because we use partial keys for iterating). I'm now leaning more towards leaving it as is.